Feature/co 1378 implement in express

wilsonia opened the pull request.

ianbrown9475 committed:
Use child_process.spawn instead of gulp-run-command

ianbrown9475 committed:
Use gulp recommended task syntax, update precommit

ianbrown9475 committed:
Add newline

wilsonia committed:
implement serializer

wilsonia committed:
implement /osuuid

wilsonia committed:
add top-level self link to design

wilsonia committed:
merge bug fix for paginator

ianbrown9475 committed:
Update readme

ianbrown9475 committed:
Use imports in app.js, use babel import resolver

ianbrown9475 committed:
Merge branch 'master' into feature/CO-1317-investigate-babel-integration

ianbrown9475 committed:
Simplify source-map-support import

ianbrown9475 committed:
Add import/order rule

ianbrown9475 committed:
ESLint sort imports alphabetically within groups

wilsonia committed:
merge bug fix for paginator

wilsonia committed:
merge paginator

wilsonia committed:
Merge remote-tracking branch 'skeleton/feature/minor-fix' into feature/CO-1378-implement-in-express

wilsonia committed:
Merge remote-tracking branch 'skeleton/feature/minor-fix' into feature/CO-1378-implement-in-express

wilsonia committed:
Merge remote-tracking branch 'skeleton/feature/minor-fix' into feature/CO-1378-implement-in-express

wilsonia committed:
implement error handling

wilsonia committed:
implement error handling for path param

wilsonia committed:
generalize ldap connection

wilsonia committed:
fix lint errors

tsoliangwu0130 committed:
Fix typo

wilsonia committed:
use forEach iterator

wilsonia committed:
config example

wilsonia committed:
Merge branch 'master' into feature/CO-1378-implement-in-express

tsoliangwu0130 commented on api/v2/db/ldap/connection.js:
```diff
@@ -0,0 +1,28 @@
+const config = require('config');
+const ldap = require('ldapjs');
+
+const ldapConfig = config.get('dataSources').ldap;
```
The better practice to get a nested config object with [config module](https://www.npmjs.com/package/config#quick-start) is like:

```js
config.get('dataSources.ldap');
```

example: https://github.com/osu-mist/notes-api/blob/master/api/v1/db/json/fs-operations.js#L4

tsoliangwu0130 commented on api/v2/db/ldap/connection.js:
```diff
@@ -0,0 +1,28 @@
+/**
+ * @summary Get an ldap connection
+ * @function
+ * @returns ldapjs client connection object
```
Can you also describe the returns type like here? https://github.com/osu-mist/notes-api/blob/master/api/v1/db/json/fs-operations.js#L30

wilsonia committed:
remove AWS from config example

tsoliangwu0130 commented on api/v2/db/ldap/connection.js:
```diff
@@ -0,0 +1,28 @@
+ * @function
+ * @returns ldapjs client connection object
+ */
+const getClient = () => ldap.createClient({ url: ldapConfig.url });
```
Looks like `ldapConfig.url` is the only field we need for creating the connection. You can use [destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) at line 4 to shorten the code:

```js
const { url } = config.get('dataSources.ldap');
```

tsoliangwu0130 reviewed (commented):
1. When I ran the API locally and send this request:

    ```
    GET https://localhost:8080/api/v2/directory
    ```

    I got a 500 error and I also noticed there is an `undefined` printed out at the console log but not sure what it was.

2. Can you add an example to all fields in [Attributes](https://github.com/osu-mist/directory-api/blob/feature/CO-1378-implement-in-express/openapi.yaml#L217)? For certain fields in openapi file like officePhoneNumber, I think we should also add format and also serialize the raw data correcspondingly. Example: https://github.com/osu-mist/persons-api/blob/master/swagger.yaml#L419

3. Naming convension: it should be `osuUid` instead of `osuuid`.

4. For the returning addresses, what does `$` mean? For example:

    ```
    "departmentMailingAddress": "Entrprse Computing Svcs$Oregon State University$3731 SW Jefferson$Corvallis, OR 97333-3971",
    ```

5. For `primaryAffiliation` in the returning data, since the [query filters are not abbrevations](https://github.com/osu-mist/directory-api/blob/feature/CO-1378-implement-in-express/openapi.yaml#L55-L60), I think it'll make more sense to return the full name in the response as well.

tsoliangwu0130 commented on api/v2/db/ldap/connection.js:
```diff
@@ -0,0 +1,28 @@
+  try {
+    const client = getClient();
+    // Perform some search that returns less than 200 results
+    client.search('o=orst.edu', { filter: 'osuPrimaryAffiliation=Retiree', scope: 'sub' }, () => {});
```
As we discussed, do you know can we find some generic query like `SELECT 1 FROM DUAL` in SQL that we can use to validate connection without depending on any real data? If not, I'm fine with your current method, but I'd suggest moving this query to config file since the LDAP schema could be changed in the future.

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+const { serializeDirectories, serializeDirectory } = require('../../serializers/directory-serializer');
+
+const { openapi } = appRoot.require('utils/load-openapi');
+const conn = appRoot.require(`api/${openapi.basePath}/db/ldap/connection`);
```
Since you're using `app-root-path` module to require ldap connection here, the eslint warning won't be triggered for dynamically importing/exporting in require statement, but I still believe it's not a good practice though. However, I can tell the why you wanna do this. Since this file is also in the `v2` folder, I'd recommend to import the connection file with relative path like [here](https://github.com/osu-mist/terms-api/blob/master/api/v1/db/oracledb/terms-dao.js#L4).

Ref: https://github.com/osu-mist/students-api/blob/master/api/v1/db/oracledb/students-dao.js#L6

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+    ['phoneNumber', 'telephoneNumber'],
+    ['officeAddress', 'osuOfficeAddress'],
+    ['department', 'osuDepartment'],
+  ]);
```
Just curious, is there any reason you decided to use a "map" instead of an "object" for key-value storing collection? I noticed you used `new Map` a lot in your code.

Ref: https://medium.com/front-end-weekly/es6-map-vs-object-what-and-when-b80621932373

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+        return fuzzyFilters;
+      }
+      case 'lastName': {
+        return `${keyMap.get(key)}=${value}*`;
```
Isn't it the same with the `firstName`?

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+        const valueTerms = value.split(/[ ,]+/);
+
+        // Loop through possible splits of query into first and last names
+        for (let commaIndex = 1; commaIndex <= valueTerms.length; commaIndex += 1) {
```
Compare with using the plain for-loop like this, since we've installed [lodash](https://lodash.com/) to the skeleton, it's pretty easy to iterate items by using [`_.forEach` (or `_.each`)](https://lodash.com/docs/4.17.11#forEach). In my opinion, this is more clean and neat and pretty similar to the Python syntax that we usually used. 

Example: https://github.com/osu-mist/terms-api/blob/master/tests/unit/terms-dao.test.js#L54-L60

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+          for (let i = 0; i < commaIndex; i += 1) firstName += `${valueTerms[i]} `;
+          for (let i = commaIndex; i < valueTerms.length; i += 1) lastName += `${valueTerms[i]} `;
+          // Consider first, last ordering and last, first ordering
+          fuzzyFilters += `(${keyMap.get(key)}=${firstName.slice(0, -1)}*, ${lastName.slice(0, -1)}*)`;
```
`keyMap.get(key)`, `firstName.slice(0, -1)` and `lastName.slice(0, -1)` are used multiple times. Can you store them to a variables and reuse them?

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+          for (let i = commaIndex; i < valueTerms.length; i += 1) lastName += `${valueTerms[i]} `;
+          // Consider first, last ordering and last, first ordering
+          fuzzyFilters += `(${keyMap.get(key)}=${firstName.slice(0, -1)}*, ${lastName.slice(0, -1)}*)`;
+          fuzzyFilters += `(${keyMap.get(key)}=${lastName.slice(0, -1)}*, ${firstName.slice(0, -1)}*)`;
```
Looks like what you were trying to do here is to remove the last item from `lastName` and `firstName`. So, if  `lastName`/`firstName` only contains one item, do we still wanna remove it?

wilsonia replied:
The purpose of using slice was to remove a trailing space character from the end of firstName and lastName. In the for loops directly above, a list of words are concatenated, with a space at the end of each one for separation.

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+        ]).get(value)}`;
+      }
+      default: {
+        return `${keyMap.get(key)}=*${value}*`;
```
`${keyMap.get(key)}=*${value}*` was also used at least twice (in `phoneNumber`). Can you also reuse it?

wilsonia committed:
handle no query parameters specified

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+  };
+
+  let ldapQuery = '(&'; // begin requiring all conditions
+  Object.keys(endpointQuery).forEach((key) => {
```
lodash also has [_.keys](https://lodash.com/docs/4.17.11#keys) which you can use.

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+  Object.keys(endpointQuery).forEach((key) => {
+    if (keyMap.get(key)) ldapQuery += `(${keyOperations(key, endpointQuery[key])})`;
+  });
+  ldapQuery += ')'; // end requiring all conditions
```
So, if `endpointQuery` is empty, the `ldapQuery` will become `(&)`. Is it expected?

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+ * @function
+ * @returns {Promise} Promise object represents a directory
+ */
+const getDirectory = pathParameter => new Promise(async (resolve, reject) => {
```
You don't need to wrap a promise twice: https://github.com/osu-mist/students-api/blob/master/api/v1/db/oracledb/students-dao.js#L18
In addition, why we need an async/await function here?

tsoliangwu0130 commented on api/v2/db/ldap/connection.js:
```diff
@@ -0,0 +1,28 @@
+ * @function
+ * @throws Throws an error if unable to connect or search ldap
+ */
+const validateLdap = async () => {
```
Why we wanna use an async/await function here? Do you await something?

wilsonia replied:
Without async, I receive the error:
```

 /Users/ian/directory-api/utils/validate-data-source.js:36
      validationMethods[dataSourceType]().catch((err) => {
                                         ^

TypeError: Cannot read property 'catch' of undefined
    at _.each (/Users/ian/directory-api/utils/validate-data-source.js:36:42)
    at arrayEach (/Users/ian/directory-api/node_modules/lodash/lodash.js:516:11)
    at Function.forEach (/Users/ian/directory-api/node_modules/lodash/lodash.js:9344:14)
    at validateDataSource (/Users/ian/directory-api/utils/validate-data-source.js:34:5)
    at Object.<anonymous> (/Users/ian/directory-api/app.js:22:1)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
```

ianbrown9475 replied:
Right, it looks like the validateDataSource method expects the validation methods to return a promise. I think it should be fine to just make `validateLdap` use async.

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+const getDirectory = pathParameter => new Promise(async (resolve, reject) => {
+  const client = conn.getClient();
+
+  client.search('o=orst.edu', { filter: `osuUID=${pathParameter}`, scope: 'sub' }, (err, res) => {
```
Looks like this query template was used multiple times as well, can we reuse it? In addition, looks like it's a callback style funciton, can we use promise instead?

ianbrown9475 replied:
I couldn't find any documentation on promise alternatives for this package but we could always use `util.promisify` https://nodejs.org/docs/latest-v10.x/api/util.html#util_util_promisify_original

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+      reject(error);
+    });
+    res.on('end', () => {
+      resolve(null);
```
To be consistent, can you return `undefined` here?

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,125 @@
+/**
+ * @summary Return a directory
+ * @function
+ * @returns {Promise} Promise object represents a directory
```
Can you specify what object will be returned if the promise is resolved?

Example: https://github.com/osu-mist/terms-api/blob/master/api/v1/db/oracledb/terms-dao.js#L14

tsoliangwu0130 commented on api/v2/paths/directory.js:
```diff
@@ -0,0 +1,28 @@
+    return res.send(result);
+  } catch (err) {
+    if ('lde_message' in err) {
+      if (err.lde_message.includes('Size Limit Exceeded')) {
```
You should be able to destructure `lde_message` from err as well:

```js
const { lde_message } = err;
if (lde_message) {
  // do something
}
```

wilsonia replied:
variable lde_message returns lint errors because it is not camel case, so I define 
`ldeMessage = err.lde_message`

ianbrown9475 replied:
It's not necessary, but you can also destructure and rename by doing something like this:
```js
const { lde_message: ldeMessage } = err;
```

tsoliangwu0130 commented on api/v2/paths/directory.js:
```diff
@@ -0,0 +1,28 @@
+      if (err.lde_message.includes('Size Limit Exceeded')) {
+        return errorBuilder(res, 400, ['Size Limit of 200 Results Exceeded (search too broad)']);
+      }
+      return errorHandler(res, [err.lde_message]);
```
Why `err.lde_message` need to be wrapped within an array?

wilsonia replied:
the skeleton expects a list of error messages in general, so if a single string is passed it is interpreted as a list of characters:
```
{
    "errors": [
        {
            "status": "400",
            "code": "1400",
            "title": "Bad Request",
            "detail": "S",
            "links": {
                "about": "https://developer.oregonstate.edu/documentation/error-reference#1400"
            }
        },
        {
            "status": "400",
            "code": "1400",
            "title": "Bad Request",
            "detail": "i",
            "links": {
                "about": "https://developer.oregonstate.edu/documentation/error-reference#1400"
            }
        },
        {
            "status": "400",
            "code": "1400",
            "title": "Bad Request",
            "detail": "z",
            "links": {
                "about": "https://developer.oregonstate.edu/documentation/error-reference#1400"
            }
        },
        {
            "status": "400",
            "code": "1400",
            "title": "Bad Request",
            "detail": "e",
            "links": {
                "about": "https://developer.oregonstate.edu/documentation/error-reference#1400"
            }
        },
        {
            "status": "400",
            "code": "1400",
            "title": "Bad Request",
            "detail": " ",
            "links": {
                "about": "https://developer.oregonstate.edu/documentation/error-reference#1400"
            }
        },
     ...
    ]
}
```

tsoliangwu0130 commented on api/v2/serializers/directory-serializer.js:
```diff
@@ -0,0 +1,129 @@
+  ['uid', 'username'],
+  ['osuAltPhoneNumber', 'alternatePhoneNumber'],
+  ['osuUID', 'osuuid'],
+]).get(attribute);
```
Same comment as above (map vs. object).

tsoliangwu0130 commented on api/v2/serializers/directory-serializer.js:
```diff
@@ -0,0 +1,129 @@
+  ['username', 'uid'],
+  ['alternatePhoneNumber', 'osuAltPhoneNumber'],
+  ['osuuid', 'osuUID'],
+]).get(attribute);
```
Same comment as above (map vs. object).

tsoliangwu0130 commented on api/v2/serializers/directory-serializer.js:
```diff
@@ -0,0 +1,129 @@
+//       resourcePath:
+//     };
+// }
+
```
What are these lines for? Can you remove them if they are not needed?

tsoliangwu0130 commented on config/default-example.yaml:
```diff
@@ -19,22 +19,6 @@ logger:
-    s3ForcePathStyle: false
+  dataSources: ['ldap']
+  ldap:
+    url: 'ldap://directory.oregonstate.edu:389'
```
We should use `ldaps`portocol. In addition, we need to make sure we are connecting to `ldap.onid.orst.edu` because that is the name in the SSL certificate. By using these, I don't think port will be necessary.

tsoliangwu0130 commented on openapi.yaml:
```diff
@@ -262,132 +250,20 @@ definitions:
         type: string
         description: HTTP status code
-<<<<<<< HEAD
         example: 123
```
Based on our style guild, if the value can be parsed differently than the expected type, or the string contains special characters, the value should be quoted.

Ref: 
* https://github.com/osu-mist/style-guides/blob/master/OpenAPI.md
* https://github.com/osu-mist/express-api-skeleton/blob/master/openapi.yaml#L193

tsoliangwu0130 commented on utils/validate-data-source.js:
```diff
@@ -2,15 +2,20 @@ const appRoot = require('app-root-path');
 const { dataSources } = config.get('dataSources');
 const json = dataSources.includes('json')
-  ? appRoot.require('api/v1/db/json/fs-operations').validateJsonDb
+  ? appRoot.require(`/api${openapi.basePath}/db/json/fs-operations`).validateJsonDb
```
When using `appRoot` require, I think we don't need the leading `/`. In addition, looks like the logic of these validation methods are pretty much the same, do you think we can write a generic function in the skeleton to share the similar logic?

tsoliangwu0130 commented on package.json:
```diff
@@ -50,7 +51,7 @@
+    "gulp": "^4.0.0",
     "gulp-eslint": "^5.0.0",
     "gulp-mocha": "^6.0.0",
     "pre-commit": "^1.2.2"
```
Some info in this file needs to be updated.

wilsonia committed:
attribute examples

ianbrown9475 committed:
move src files to src/ directory

ianbrown9475 committed:
add gulpfile comments

ianbrown9475 committed:
simplify mock-data import

ianbrown9475 committed:
WIP convert require() to import

wilsonia committed:
value operations

wilsonia committed:
improve ldap connection

wilsonia committed:
map -> object

ianbrown9475 committed:
WIP fix json db path, express-openapi paths

ianbrown9475 committed:
Merge branch 'master' into feature/CO-1317-investigate-babel-integration

ianbrown9475 committed:
Move compiled files to dist/, fix import errors

tsoliangwu0130 committed:
Add tuple types for float, double, and number

wilsonia committed:
simplify logic for fuzzyName

tsoliangwu0130 committed:
Fix bug

tsoliangwu0130 committed:
Add items check

tsoliangwu0130 committed:
Fix bug

ianbrown9475 committed:
Fix eslint import ignore settings

tsoliangwu0130 committed:
Merge pull request #30 from osu-mist/feature/fix-typo

Fix typo

ianbrown9475 committed:
Move .babelrc -> babel.config.js, fix .gitignore, add node versions

ianbrown9475 committed:
Use null safe operator, use current node version in babel env

ianbrown9475 committed:
Merge branch 'master' into feature/CO-1317-investigate-babel-integration

tsoliangwu0130 committed:
Clean code

tsoliangwu0130 committed:
Update style

tsoliangwu0130 committed:
Fix W605

tsoliangwu0130 committed:
Merge pull request #32 from osu-mist/feature/CO-1397-students-api-integration-test

CO-1397 students api integration test

wilsonia committed:
improve empty query handling

wilsonia committed:
fix lint errors

wilsonia committed:
minor issues

wilsonia committed:
map -> object

ianbrown9475 committed:
Add better message in make_request

(cherry picked from commit 6bad403695eb9ea833379cd2e06d94878d98e84e)

tsoliangwu0130 committed:
Adhere style guide

tsoliangwu0130 committed:
Update integration test

wilsonia committed:
use range in for loop

wilsonia committed:
examples in ""

tsoliangwu0130 committed:
Bump express-basic-auth from 1.1.5 to 1.1.7

tsoliangwu0130 committed:
Merge pull request #35 from osu-mist/feature/CO-1461-fix-security-vulnerability

CO-1461 Bump express-basic-auth from 1.1.5 to 1.1.7

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,126 @@
+
+        return fuzzyFilters;
+      }
+      case 'firstName':
```
What do these cases do? Won't they go into the `default` case?

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,126 @@
+
+          // Add leading wildcard for non-null first and last names
+          firstName = (firstName ? `*${firstName}` : firstName);
+          lastName = (lastName ? `*${lastName}` : lastName);
```
Parentheses are not needed.

tsoliangwu0130 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,126 @@
+
+          // Consider first, last ordering and last, first ordering with trailing wildcard
+          fuzzyFilters += `(${keyMap[key]}=${firstName}*, ${lastName}*)`;
+          fuzzyFilters += `(${keyMap[key]}=${lastName}*, ${firstName}*)`;
```
`keyMap[key]` is used many times in this block. Can we store it into a var and reuse it?

tsoliangwu0130 commented on api/v2/serializers/directory-serializer.js:
```diff
@@ -0,0 +1,148 @@
+  username: 'uid',
+  alternatePhoneNumber: 'osuAltPhoneNumber',
+  osuUid: 'osuUID',
+}[attribute]);
```
I guess you can use lodash to improve this. For example:

```js
const _ = require('lodash');

const lower_to_upper = {
  'a': 'A',
  'b': 'B'
};

const upper_to_lower = {};

_.each(lower_to_upper, (v, k) => {
  upper_to_lower[v] = k;
});
```

ianbrown9475 replied:
I think you can make this even simpler using invert https://lodash.com/docs/#invert
```js
const _ = require('lodash');

const lower_to_upper = {
  'a': 'A',
  'b': 'B',
};
const upper_to_lower = _.invert(lower_to_upper);
// { 'A': 'a', 'B': 'b' }
```

tsoliangwu0130 commented on api/v2/serializers/directory-serializer.js:
```diff
@@ -0,0 +1,148 @@
+        O: 'Other',
+        R: 'Retiree',
+        U: 'Unknown',
+      }[value];
```
Looks like the map here is pretty much the same as what you did in DAO object expect they are reversed. Can you figure a way to maintain only a map in one place?

ianbrown9475 commented on api/v2/db/ldap/connection.js:
```diff
@@ -0,0 +1,24 @@
+ */
+const validateLdap = async () => {
+  ldap.createClient({ url }).on('error', () => {
+    throw new Error('Error connecting to ldap');
```
It looks like `createClient` will only emit an `error` event if you pass in `bindDN` and `bindCredentials` which will cause an implicit bind operation https://github.com/joyent/node-ldapjs/blob/ad451edc18d7768c3ddee1a1dd472d2bbafdae5e/lib/client/client.js#L346. Since that's not happening, this shouldn't be triggered. Also exceptions may be thrown here if the url is invalid for example so the call here should be wrapped in a try-catch block if you want to throw a different error.

wilsonia replied:
`createClient` could also emit an `error` event on a failed attempt at reconnecting:
https://github.com/joyent/node-ldapjs/blob/ad451edc18d7768c3ddee1a1dd472d2bbafdae5e/lib/client/client.js#L1228
https://github.com/joyent/node-ldapjs/blob/ad451edc18d7768c3ddee1a1dd472d2bbafdae5e/lib/client/client.js#L1290

```
Error: Error connecting to ldap
    at Client.<anonymous> (/home/ian/directory-api/api/v2/db/ldap/connection.js:20:11)
    at Client.emit (events.js:200:13)
    at Backoff.<anonymous> (/home/ian/directory-api/node_modules/ldapjs/lib/client/client.js:1228:12)
    at Backoff.emit (events.js:200:13)
    at Backoff.backoff (/home/ian/directory-api/node_modules/backoff/lib/backoff.js:41:14)
    at /home/ian/directory-api/node_modules/ldapjs/lib/client/client.js:1214:15
    at f (/home/ian/directory-api/node_modules/once/once.js:25:25)
    at TLSSocket.onResult (/home/ian/directory-api/node_modules/ldapjs/lib/client/client.js:1016:7)
    at Object.onceWrapper (events.js:288:20)
    at TLSSocket.emit (events.js:200:13)
```

This is the error that is omitted when an invalid url is used in the configuration. When wrapping with a try / catch block, the exception is not actually caught.
```
Error: getaddrinfo ENOTFOUND ldap.fsdafsdonid.orst.edu
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:60:26)
Emitted 'error' event at:
    at Backoff.<anonymous> (/home/ian/directory-api/node_modules/ldapjs/lib/client/client.js:1228:12)
    at Backoff.emit (events.js:200:13)
    at Backoff.backoff (/home/ian/directory-api/node_modules/backoff/lib/backoff.js:41:14)
    at /home/ian/directory-api/node_modules/ldapjs/lib/client/client.js:1214:15
    at f (/home/ian/directory-api/node_modules/once/once.js:25:25)
    at TLSSocket.onResult (/home/ian/directory-api/node_modules/ldapjs/lib/client/client.js:1016:7)
    at Object.onceWrapper (events.js:288:20)
    at TLSSocket.emit (events.js:200:13)
    at emitErrorNT (internal/streams/destroy.js:91:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:59:3) {
  errno: 'ENOTFOUND',
  code: 'ENOTFOUND',
  syscall: 'getaddrinfo',
  hostname: 'ldap.fsdafsdonid.orst.edu'
}
events.js:177
      throw er; // Unhandled 'error' event
```

ianbrown9475 replied:
I tried the url `ldap.fsdafsdonid.orst.edu` and got this exception:
```
TypeError: ldap.fsdafsdonid.orst.edu is an invalid LDAP url (protocol)
    at Object.parse (/Users/ian/github/directory-api/node_modules/ldapjs/lib/url.js:16:13)
    at new Client (/Users/ian/github/directory-api/node_modules/ldapjs/lib/client/client.js:310:16)
    at Object.createClient (/Users/ian/github/directory-api/node_modules/ldapjs/lib/client/index.js:54:12)
    at Object.validateLdap [as ldap] (/Users/ian/github/directory-api/api/v2/db/ldap/connection.js:19:23)
    at _.each (/Users/ian/github/directory-api/utils/validate-data-source.js:36:40)
    at arrayEach (/Users/ian/github/directory-api/node_modules/lodash/lodash.js:516:11)
    at Function.forEach (/Users/ian/github/directory-api/node_modules/lodash/lodash.js:9344:14)
    at validateDataSource (/Users/ian/github/directory-api/utils/validate-data-source.js:34:5)
    at Object.<anonymous> (/Users/ian/github/directory-api/app.js:22:1)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
```
There appears to be at least one case where an exception may be thrown. It looks like there's documentation on this too: https://github.com/joyent/node-ldapjs/blob/ad451edc18d7768c3ddee1a1dd472d2bbafdae5e/lib/client/client.js#L300

ianbrown9475 commented on api/v2/db/ldap/connection.js:
```diff
@@ -0,0 +1,24 @@
+ * @throws Throws an error if unable to connect or search ldap
+ */
+const validateLdap = async () => {
+  ldap.createClient({ url }).on('error', () => {
```
If we're creating a client here just for validation purposes, we should probably destroy it immediately after. There isn't very good documentation on this but it looks like there is a `client.destroy()` method that could handle this https://github.com/joyent/node-ldapjs/issues/382.

ianbrown9475 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,126 @@
+  _.keys(endpointQuery).forEach((key) => {
+    if (keyMap[key]) ldapQuery += `(${valueOperations(key, endpointQuery[key])})`;
+  });
+  if (ldapQuery) ldapQuery = `(&${ldapQuery})`;
```
Just my opinion, but I've always tried to avoid omitting the braces for if-else blocks. You could also use the ternary operator for these statements.

ianbrown9475 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,126 @@
+/**
+ * @summary Return a list of directories
+ * @function
+ * @returns {Promise<object>} Promise object represents a serilized list of directory resrouces
```
Typo serilized -> serialized

ianbrown9475 commented on api/v2/paths/directory.js:
```diff
@@ -0,0 +1,32 @@
+    const ldeMessage = err.lde_message;
+    if (ldeMessage) {
+      if (ldeMessage.includes('Size Limit Exceeded')) {
+        return errorBuilder(res, 400, ['Size Limit of 200 Results Exceeded (search too broad)']);
```
What's this 200 size limit set by? Do we want to specify this if it's possible that it could change in the future or just respond with a generic 'Size limit exceeded' error?

ianbrown9475 commented on openapi.yaml:
```diff
@@ -128,89 +125,80 @@ paths:
           description: Not found
           schema:
             $ref: "#/definitions/Errors"
-=======
```
Could you also change `osuuid` to `osuUid` in the rest of the openapi document?

ianbrown9475 reviewed (changes requested):
Looks like the API is working well. Nice work! From our conversation, we should decide if the `onid` filter parameter should be fuzzy at the end. Also, when querying a directory by its `osuUid`, the value of `primaryAffiliation` is the one-character abbreviation instead of the entire word. I don't know if this is intentional or not but I would assume that it should be the full word to be consistent with the API document.

ianbrown9475 committed:
Check for unexpected fields in check_schema

(cherry picked from commit b4f4fee766d4079a72a6bdef2d2014d616486cf4)

wilsonia committed:
dao improvements

wilsonia committed:
use one inverted map

wilsonia committed:
use one inverted map

wilsonia committed:
minor fixes

wilsonia committed:
destory validation client

tsoliangwu0130 committed:
Close connection after validation is done

tsoliangwu0130 committed:
Add error message

tsoliangwu0130 committed:
Merge pull request #36 from osu-mist/feature/hot-fix

Close connection after validation is done

wilsonia committed:
remove size limit details (could change)

wilsonia committed:
promisify

ianbrown9475 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,129 @@
+        const valueTerms = value.split(/[ ,]+/);
+        _.forEach(_.range(valueTerms.length), (commaIndex) => {
+          let firstName = `${valueTerms.slice(0, commaIndex).join(' ')}`;
+          let lastName = `${valueTerms.slice(commaIndex).join(' ')}`;
```
I think you could remove the ` `` ` and `${}` here

ianbrown9475 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,129 @@
+  let ldapQuery = '';
+  _.keys(endpointQuery).forEach((key) => {
+    if (keyMap[key]) ldapQuery += `(${valueOperations(key, endpointQuery[key])})`;
+  });
```
Not necessary, but you could also do something like
```js
_.forEach(endpointQuery, (value, key) => {
  if (keyMap[key]) ldapQuery += `(${valueOperations(key, value)})`;
});
```

ianbrown9475 commented on api/v2/db/ldap/directory-dao.js:
```diff
@@ -0,0 +1,129 @@
+/**
+ * @summary Return a list of directories
+ * @function
+ * @returns {Promise<object>} Promise object represents a serialized list of directory resrouces
```
Typo resrouces -> resources.
Could you also write something about how `undefined` is returned if the query is empty?

ianbrown9475 commented on api/v2/paths/directory/{osuUid}.js:
```diff
@@ -10,10 +10,10 @@ const { openapi: { paths } } = appRoot.require('utils/load-openapi');
  */
```
Could you update this documentation?

ianbrown9475 reviewed (changes requested):
Could you also update the [readme](readme.md), [Dockerfile](Dockerfile) and [package.json](package.json)?

ianbrown9475 committed:
Don't ignore query parameters with empty values

(cherry picked from commit 4766c8af4d47bb5a590adebb64638cbe3aaeefe1)

ianbrown9475 committed:
Fix quoting style

(cherry picked from commit dc129dee93efbf239d8587c93cbac67512fb2337)

ianbrown9475 committed:
Merge pull request #37 from osu-mist/feature/pull-notes-api-integration-test-changes

Pull notes api integration test changes

tsoliangwu0130 committed:
User the right term

tsoliangwu0130 committed:
Merge pull request #34 from osu-mist/feature/CO-1455-adhere-style-guilde

CO-1455 Adhere style guide

Ruefa committed:
Check that err has a stack before logging the error.

Ruefa committed:
fixed format of if

wilsonia committed:
catch invalid ldap errors

wilsonia committed:
update documentation

wilsonia committed:
retian elements of skeleton README relevant to directory-api

wilsonia committed:
document fistName param

wilsonia committed:
test all pages for correct size

wilsonia committed:
test for duplicate results

wilsonia committed:
remove duplicate README

ianbrown9475 reviewed (changes requested):
Could you update `package.json` by removing references to the express skeleton?
Could you also remove parts of the readme that reference basing a new api off of the skeleton?

wilsonia committed:
write function for repeating for all pages

dependabot[bot] committed:
Bump lodash.merge from 4.6.1 to 4.6.2

Bumps [lodash.merge](https://github.com/lodash/lodash) from 4.6.1 to 4.6.2.
- [Release notes](https://github.com/lodash/lodash/releases)
- [Commits](https://github.com/lodash/lodash/commits)

Signed-off-by: dependabot[bot] <support@github.com>

tsoliangwu0130 committed:
Merge pull request #40 from osu-mist/dependabot/npm_and_yarn/lodash.merge-4.6.2

Bump lodash.merge from 4.6.1 to 4.6.2

dependabot[bot] committed:
Bump lodash from 4.17.11 to 4.17.13

Bumps [lodash](https://github.com/lodash/lodash) from 4.17.11 to 4.17.13.
- [Release notes](https://github.com/lodash/lodash/releases)
- [Commits](https://github.com/lodash/lodash/compare/4.17.11...4.17.13)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump js-yaml from 3.12.0 to 3.13.1

Bumps [js-yaml](https://github.com/nodeca/js-yaml) from 3.12.0 to 3.13.1.
- [Release notes](https://github.com/nodeca/js-yaml/releases)
- [Changelog](https://github.com/nodeca/js-yaml/blob/master/CHANGELOG.md)
- [Commits](https://github.com/nodeca/js-yaml/compare/3.12.0...3.13.1)

Signed-off-by: dependabot[bot] <support@github.com>

tsoliangwu0130 committed:
Merge pull request #41 from osu-mist/dependabot/npm_and_yarn/lodash-4.17.13

Bump lodash from 4.17.11 to 4.17.13

tsoliangwu0130 committed:
Merge pull request #42 from osu-mist/dependabot/npm_and_yarn/js-yaml-3.13.1

Bump js-yaml from 3.12.0 to 3.13.1

Ruefa committed:
simplify if statement for printing errors

Ruefa committed:
Merge pull request #38 from osu-mist/feature/CO-1499-improve-error-reporting

CO-1499 improve error reporting

wilsonia committed:
remove references to skeleton

ianbrown9475 committed:
Properly terminate execution flow after calling next()

ianbrown9475 commented on README.md:
```diff
@@ -86,30 +88,7 @@ $ gulp test
 $ npm test
```
Do you think Prerequisites step 3 should be removed since it's describing creating a new API? We might want to move this to a different section in the skeleton readme also since it's a little different than the other steps in this section

ianbrown9475 committed:
Use return in middleware functions

ianbrown9475 committed:
Add back blank line

ianbrown9475 committed:
Merge pull request #43 from osu-mist/fix/fix-runtime-errors

CO-1514: Fix runtime errors

wilsonia committed:
remove 'design API' step from README

wilsonia committed:
total pages remains constant

wilsonia committed:
page number validation

wilsonia committed:
openapi version 4.6.5

wilsonia committed:
for loop for testing all pages

wilsonia committed:
revert eslint changes

wilsonia committed:
Merge pull request #45 from osu-mist/hotfix-update-openapi

openapi version 4.6.5

wilsonia committed:
Merge branch 'master' of github.com:osu-mist/express-api-skeleton into feature/CO-1378-implement-in-express

wilsonia committed:
ErrorResult -> Errors

wilsonia committed:
Errors -> ErrorResult

ianbrown9475 committed:
Instantiate a winston logger, improve logger behavior

ianbrown9475 committed:
Turn on eslint no-console, replace console with logger

ianbrown9475 committed:
Implement custom formatter, remove duplicated properties

ianbrown9475 committed:
Log request and response body

ianbrown9475 committed:
Fix comment syntax

ianbrown9475 committed:
Set promiseMode to true

ianbrown9475 committed:
Use jsdoc eslint, update documentation

ianbrown9475 committed:
Improve logger documentation

ianbrown9475 committed:
Don't log response body or remove empty objects, fix error handler

ianbrown9475 committed:
Add comment for express types

wilsonia committed:
organize gulp tasks

wilsonia committed:
organize gulp tasks

wilsonia committed:
allow minor updates to gulp

wilsonia committed:
require gulp-cli as dev dependency

wilsonia committed:
remove gulp-cli as dev dependency

wilsonia committed:
update express

ianbrown9475 committed:
Move the logger middleware to middlwares/

ianbrown9475 committed:
Fix comment in middlewares/logger

ianbrown9475 committed:
Merge pull request #48 from osu-mist/hotfix/use-promiseMode

Set promiseMode to true

ianbrown9475 committed:
Merge pull request #47 from osu-mist/feature/CO-1519-improve-logging

CO-1519: improve logging

ianbrown9475 committed:
Remove decamelize/capitalize logic

ianbrown9475 committed:
Merge remote-tracking branch 'origin/master' into feature/CO-1526-update-jsdoc

ianbrown9475 committed:
Fix app.js logger middleware import

ianbrown9475 committed:
Merge pull request #50 from osu-mist/feature/CO-1519-improve-logging

Fix app.js logger middleware import

ianbrown9475 committed:
Change api logger level blue -> cyan

ianbrown9475 committed:
Merge pull request #49 from osu-mist/feature/CO-1526-update-jsdoc

CO-1526: Update JSDoc

ianbrown9475 committed:
Merge pull request #51 from osu-mist/feature/CO-1519-improve-logging

Change api logger level blue -> cyan

ianbrown9475 committed:
Add 409 response to runtime errors

ianbrown9475 committed:
Require jsdoc for exported arrow functions

ianbrown9475 committed:
Merge pull request #53 from osu-mist/fix/require-jsdoc-for-public-functions

Fix: Require jsdoc for exported arrow functions

ianbrown9475 committed:
Rename enumMessage -> enumErrorMessage

ianbrown9475 committed:
Merge pull request #52 from osu-mist/feature/CO-1536-use-409-for-invalid-type

CO-1536: add 409 response to runtime errors

Ruefa committed:
created function to generate invalid path parameter format error messages

Ruefa committed:
added checking and error handling for pattern.openapi.validation errors

Ruefa committed:
added check that error is in path for 404 path pattern validation errors

Ruefa committed:
removed method to generate error string and placed the generation in the error builder function call

Ruefa committed:
removed duplicate apiDoc setting and as a result removed the requirement of openapi in the paths effected.

Ruefa committed:
added pattern for petId parameter in GET /pets/{petId}

Ruefa committed:
fixed pattern on petId

Ruefa committed:
Merge pull request #55 from osu-mist/feature/CO-1537-remove-apiDoc-duplication

CO-1537 Removed duplicate apiDoc declaration

Ruefa committed:
Merge pull request #54 from osu-mist/feature/CO-1539-path-parameter-pattern-return-404

CO-1539 path parameter pattern return 404

wilsonia committed:
Merge branch 'master' into feature/CO-1424-add-pagination-unit-tests

wilsonia committed:
add jsdoc describing repeating function

wilsonia committed:
for -> while

wilsonia committed:
array -> set

wilsonia committed:
add more mock data

wilsonia committed:
test for all possible page sizes, numbers

wilsonia committed:
update yarn.lock

wilsonia committed:
split gulpfile

ianbrown9475 committed:
Add bucketExists tests

wilsonia committed:
Merge branch 'master' of github.com:osu-mist/express-api-skeleton into feature/CO-1378-implement-in-express

wilsonia committed:
resolve lint warnings

wilsonia committed:
mock data in order

wilsonia committed:
lodash simplify

wilsonia committed:
document why set is being cleared

wilsonia committed:
capitalize Axolotl

wilsonia committed:
use yarn start instead of prod gulpfile

ianbrown9475 committed:
Refactor s3 stub, add rejected bucketExists case

ianbrown9475 committed:
Add validateAwsS3 test case, use async/await in bucketExists

wilsonia committed:
Merge branch 'master' into feature/CO-1317-investigate-babel-integration

ianbrown9475 committed:
Use async/await for aws-operations functions

ianbrown9475 committed:
Add objectExists tests

ianbrown9475 committed:
Refactor aws-operations method test cases

ianbrown9475 committed:
Change parameter of createS3Stub to object

ianbrown9475 committed:
Refactor and add headObject test cases

wilsonia committed:
JSDoc

ianbrown9475 committed:
Undo refactoring, add more assertions

ianbrown9475 committed:
Fix messages

kalcicd committed:
Update Instant Client install instructions

kalcicd committed:
Merge pull request #56 from osu-mist/feature/CO-1558-update-installation-instructions-in-readme

Update Instant Client install instructions

ianbrown9475 reviewed (changes requested):
The changes look good and everything seems to work, but the current lockfile doesn't satisfy `package.json`, so installing will rebuild it. Could you rebuild the lockfile and commit the changes?

ianbrown9475 committed:
Add listObjects cases

ianbrown9475 committed:
Add getObject test cases

ianbrown9475 committed:
Add putDir and putObject test cases

ianbrown9475 committed:
Use spread syntax instead of Object.assign

ianbrown9475 committed:
Use spread syntax in app.js

ianbrown9475 committed:
Add updateMetadata test cases

ianbrown9475 committed:
Add deleteObject test cases

ianbrown9475 committed:
Add documentation

ianbrown9475 committed:
Move jsdoc eslint plugin to devDependencies

ianbrown9475 committed:
Remove extra blank line

wilsonia committed:
run forever in foreground

wilsonia committed:
add babel runtime to prod dependencies

wilsonia committed:
add gulp to prod dependencies

wilsonia committed:
move eslint to dev dep

wilsonia committed:
rebuild lockfile

wilsonia committed:
move gulp to dev dep

wilsonia committed:
move /api under /src

wilsonia committed:
use import

wilsonia committed:
merge babel integration

wilsonia committed:
merge babel integration

ianbrown9475 reviewed (changes requested):
Looks good, but could you delete the `api/` directory since this was moved to `src/api/`?

wilsonia committed:
switch from yarn to npm

wilsonia committed:
remove references to yarn

wilsonia committed:
remove unused copy of /api

ianbrown9475 committed:
Refactor error handling logic

ianbrown9475 committed:
Don't export setBucket

wilsonia committed:
add npm build script

wilsonia committed:
Merge pull request #46 from osu-mist/feature/CO-1424-add-pagination-unit-tests

Feature/co 1424 add pagination unit tests

ianbrown9475 committed:
Merge pull request #57 from osu-mist/feature/CO-1409-add-aws-unit-tests

CO-1409: add aws unit tests

wilsonia committed:
merge master

wilsonia committed:
merge master

wilsonia committed:
Merge branch 'feature/CO-1317-investigate-babel-integration' of https://github.com/osu-mist/express-api-skeleton into feature/CO-1317-investigate-babel-integration

wilsonia committed:
Merge branch 'master' into feature/CO-1317-investigate-babel-integration

wilsonia committed:
merge master

wilsonia committed:
sort package json

wilsonia committed:
Merge pull request #27 from osu-mist/feature/CO-1317-investigate-babel-integration

CO-1317: investigate babel integration

ianbrown9475 committed:
Fix aws-operations

ianbrown9475 committed:
Merge pull request #61 from osu-mist/hotfix/fix-aws-operations

Hotfix - Fix aws-operations

wilsonia committed:
allow fs-operations, aws-operations unresolved

wilsonia committed:
allow unresolved connection module

wilsonia committed:
Merge pull request #62 from osu-mist/hotfix/allow-unresolve-operations

allow fs-operations, aws-operations unresolved

wilsonia committed:
use import/export, new logger module

ianbrown9475 committed:
Remove default exports

wilsonia committed:
use import-export

ianbrown9475 committed:
Merge pull request #65 from osu-mist/hotfix/remove-default-exports

Hotfix - Remove default exports

wilsonia committed:
Merge pull request #64 from osu-mist/hotfix/update-oracle-connection

use import/export, new logger module

tsoliangwu0130 committed:
Fix path in readme

tsoliangwu0130 committed:
Fix package-lock.json

tsoliangwu0130 committed:
Fix package-lock.json

wilsonia committed:
use import

tsoliangwu0130 committed:
Add npm badge

tsoliangwu0130 committed:
Merge pull request #66 from osu-mist/hot-fix/fix-path-for-integration-test

Hot fix/fix path for integration test

ianbrown9475 committed:
Treat macOS docker localhost macro as localhost in integration tests

ianbrown9475 committed:
Fix root path, add package.json alias, remove app-root-path

ianbrown9475 committed:
Update airbnb eslint and adapt code

ianbrown9475 committed:
Update packages

ianbrown9475 committed:
Enable sourcemaps for unit tests

ianbrown9475 committed:
Update README, remove Dockerfile

ianbrown9475 committed:
Fix oracledb connection bug

ianbrown9475 committed:
Ignore extraneous dependencies in all test js files

ianbrown9475 committed:
Fix import in middlewares/authentication

ianbrown9475 committed:
Fix oracledb pool creation issue

config 3.0 now makes the configuration an immutable object. since
oracledb will modify the poolAttrs object passed to createPool, we need
to pass a mutable object instead. To fix this, a copy of the config
object is passed that only uses a selection of attributes from the
configuration object.

ianbrown9475 committed:
Import contrib example expecting named imports

ianbrown9475 committed:
Update dependencies 2019-08-26

ianbrown9475 committed:
Merge pull request #67 from osu-mist/feature/CO-1577-improve-skeleton-after-babel

Feature - CO-1577: improve skeleton after using babel

ianbrown9475 committed:
Fix contrib import

ianbrown9475 committed:
Merge pull request #68 from osu-mist/hotfix/fix-contrib-import

Hotfix - Fix contrib import

wilsonia committed:
Merge branch 'master' of github.com:osu-mist/express-api-skeleton into feature/CO-1378-implement-in-express
